[JIT] Add AliasDb API For Changing Aliasing

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31800 [JIT] check for object equality in constant pooling
* **#31501 [JIT] Add AliasDb API For Changing Aliasing**

We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

Differential Revision: [D19254413](https://our.internmc.facebook.com/intern/diff/D19254413)

Commit:
Add safeToChangeAliasingRelationship api

[ghstack-poisoned]

Commit:
Update on "[JIT] Add AliasDb API For Changing Aliasing"


We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 1da8b3e:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=1da8b3e4a11169019cd4ccc60db5d4400412a584).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4023306) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77816534) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77816534))
<details>
<summary>
<code>Dec 19 23:03:48 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff</code>
</summary>

```
Dec 19 23:03:47 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Dec 19 23:03:48 1 out of 1 hunk FAILED 
Dec 19 23:03:48 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Dec 19 23:03:48 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Dec 19 23:03:48 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Dec 19 23:03:48 Traceback (most recent call last): 
Dec 19 23:03:48   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Dec 19 23:03:48     patch_repo(args) 
Dec 19 23:03:48   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Dec 19 23:03:48     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Dec 19 23:03:48 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Dec 19 23:03:48 + cleanup 
Dec 19 23:03:48 + retcode=1 
Dec 19 23:03:48 + set +x 
Dec 19 23:03:48 =================== sccache compilation log =================== 
Dec 19 23:03:48 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 19 23:03:48 Compile requests                 1 
Dec 19 23:03:48 Compile requests executed        0 
Dec 19 23:03:48 Cache hits                       0 
Dec 19 23:03:48 Cache misses                     0 
Dec 19 23:03:48 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Add AliasDb API For Changing Aliasing"


We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -930,6 +931,36 @@ bool AliasDb::couldMoveBeforeTopologically(Node* n, Node* movePoint) {
+// cannot escape the current graph scope. Values can escape the current scope
+// by aliasing a graph output or input, or aliasing an input to
+// prim::FunctionCall or MethodCall. FunctionCall and MethodCall are marked as
+// mutating ops, so we do not need to check for them because they are covered
```
I forgot to mention `prim::SetAttr` but it follows under the same case as `FunctionCall`. 

I should probably add handling of `prim::GetAttr` as well...

zdevito replied:
One thing to be wary of here is that `hasWriters` doesn't express the intent of this query, so while it may be  correct for todays definition, someone might change hasWriters to mean "does this alias something, and do I see a write to it", without realizing it is being used here to ask if a value escapes. It might be more accurate to query whether the value aliases the heap location directly.

eellison replied:
Updated

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -930,6 +931,36 @@ bool AliasDb::couldMoveBeforeTopologically(Node* n, Node* movePoint) {
+  bool a_escapes_scope =
+      mayContainAlias(graph_inputs, a) || mayContainAlias(graph_outputs, b);
+  bool b_escapes_scope =
+      mayContainAlias(graph_inputs, a) || mayContainAlias(graph_outputs, b);
```
both these bools compute the same thing, seems like a copy/paste error. But if it doesn't fail any tests, it also indicates this function is pretty undertested.

Commit:
Update on "[JIT] Add AliasDb API For Changing Aliasing"


We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

[ghstack-poisoned]

eellison reviewed (commented):
Updated to change the condition so that it's:
- a and b both can't have writers
- both a and b can't escape scope

- escape scope by aliasing a graph input or output, or by aliasing the wildcard set. 

It's a little unfortunate that being used in a list causes a value to alias the wildcard set, but oh well.

zdevito reviewed (approved):
Cool. My thought on the long term for how we handle this stuff is to locate almost all of the aliasing logic inside a pass that produces functional torchscript (i.e. either subgraphs that are truly functional, or inserts type-level distinction between mutable/non-mutable tensors). That way CSE could just work on the functional operators.

Commit:
Update on "[JIT] Add AliasDb API For Changing Aliasing"


We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

Differential Revision: [D19254413](https://our.internmc.facebook.com/intern/diff/D19254413)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add AliasDb API For Changing Aliasing"


We have a number of places in our code base where we should be checking if it's safe to change the alias relationship between two sets of values. This PR adds an api to Alias Db to consolidate the logic, and refactors Constant Pooling and `CSE` to use the new api. Next steps: add api usage in peephole.cpp where applicable. 

Happy to bikeshed `AliasDb::safeToChangeAliasingRelationship`. Previously I suggested `AliasDb::safeToIntroduceAliasing`, however that's not quite accurate, because this API also handles when it is unsafe to remove aliasing. 

Alternate suggestions: `safeToChangeAliasing`, `validToChangeAliasing`, `validToChangeAliasingRelationship`

Related:  https://github.com/pytorch/pytorch/issues/28360

Differential Revision: [D19254413](https://our.internmc.facebook.com/intern/diff/D19254413)

[ghstack-poisoned]
